src/main/java/seedu/address/model/ModelManager.java

```diff
@@ -79,6 +88,49 @@ public synchronized void addTask(Task person) {
         indicateTaskBookChanged();
     }
 
+    //=============for undo and redo===================================
+    @Override
+    public Command undo() throws IllegalValueException {
+
+        if (head <= 0) {
+            System.out.println(head);
+            throw new IllegalValueException("No actions to undo");
+        }
+        head--;
+        Commit commit = commits.get(head);
+        resetData(new TaskBook(commit.getTaskBook()));
+        return commit.getCommand();
+    }
+
+    @Override
+    public void recordState(Command command) {
+        commits.add(new Commit(command, new TaskBook(getTaskBook())));
+        this.head = commits.size() - 1;
+    }
+
+    @Override
+    public Command redo() throws IllegalValueException {
+        if (head >= commits.size() - 1) {
+            System.out.println(head);
+            throw new IllegalValueException("no undos to redo");
+        }
+        head++;
+        Commit commit = commits.get(head);
+        resetData(new TaskBook(commit.getTaskBook()));
+        return commit.getCommand();
+    }
+
+    /**
+     * check if taskBook has changed.
+     * @return
+     */
+    public boolean hasUncommittedChanges() throws EmptyStackException {
+        if (commits.isEmpty()) {
+            return true;
+        }
+        return !(this.taskBook.equals(commits.get(commits.size() - 1).getTaskBook()));
```

Comment:
What happens when I run

1. `recordState()`
2. `recordState()`
3. `undo()`
4. `hasUncommittedChanges()`?

Which commit will be used for comparison? I think there is a bug here.